Added validate command

sorenmat opened the pull request:
Added verify command to validate .tf files.

This can be usefull if you want to verify your terraform files on a CI server or as a webhook for
verifying commits to a repository.
It's a simpler verfication than running a terraform plan.
It will verify:
 * Missing or undefined variables
 * HCL structure error i.e. wrongly named/missing references

sorenmat commented:
@radeksimko Any decision on this one ?

radeksimko commented:
Hey @sorenmat ,
I really like the idea. I was myself looking for something like this in our CI, I just didn't have time to review the code thoroughly.

Also it would help the reviewer & users if you attached docs describing the purpose of that command - i.e. what are the typical errors/mistakes that _will_ be detected and _won't_ be detected.

e.g. circular dependencies, offline validation (`ValidateFunc` in schema), invalid references, invalid HCL syntax, ... ?

-------

It's not quite solving https://github.com/hashicorp/terraform/issues/3190 yet (ain't saying it should), but it may in the future.

sorenmat commented:
I think it's pretty close to solving #3190 from how I read it. Atleast that is how we use it.
We run it on our CI servers to validate the .tf files for syntax errors.
So it kinda works like plan in offline mode.

josephholsten commented:
@sorenmat this seems like a no-brainer. Got time to update this onto master? If not, I'll gladly do it.

@phinze, @mitchellh this could use your :eyes". Has tests, needs doc & rebase.

josephholsten commented:
lol, the only merge conflict was in `.gitignore`. rebased up in #5004

sorenmat commented:
@josephholsten I rebased an added some documentation to website/source/docs/commands
If someone could proof read it, that would be awesome.

josephholsten commented:
looks great!

@phinze this is fully ready for your :eyes:, has tests & doc.

radeksimko commented:
Functionally this looks great.

There's a few things I'd like to discuss though:

 - the term `verify` is often used in the context of formal (e.g. by certification authority) verification which ends with getting a certificate. I'd say that `validate` is more precise term for what we're doing here. After all most of the underlying method names are also using this term. Would you mind renaming the command?
 - The test case you included is great, however would you mind also adding a few more which would cover other failures? It doesn't have to be covering all the cases, although it would be awesome :trophy:  , but 2-3 more would be nice. See my list below. :arrow_down: 
 - I think we should mention some examples in the documentation (i.e. what is considered a failure) :arrow_down: 

```md
 * invalid [HCL](https://github.com/hashicorp/hcl) syntax (e.g. missing trailing quote or equal sign)
 * invalid HCL references (e.g. variable name or attribute which doesn't exist)
 * same `provider` declared multiple times
 * same `module` declared multiple times
 * same `resource` declared multiple times
 * same `output` declared multiple times
 * invalid `module` name
 * interpolation used in places where it's unsupported
```

radeksimko commented on website/source/docs/commands/verify.html.markdown:
```diff
@@ -0,0 +1,20 @@
+
+The `terraform verify` command is used to validate the format and structure of the terraform files.
+Terraform performas an analysis of the terraform files, and will display information about missing
+ variables and errors in the structure of the files.
```
Can we reword this a little bit to not mention `format` and `structure`? What we're really checking is _syntax_ mostly.

Formatting will be checked by `fmt` per https://github.com/hashicorp/terraform/pull/4955 eventually.

I would in fact even explicitly mentioned in a separate paragraph that this command does not check the format of your code. We may expand that paragraph later on when we merge `fmt` as linked above.

------------

It's pretty much like `go vet` VS `gofmt` whereas `go vet` is what we do.

radeksimko commented:
Besides the things mentioned, this looks really good, I'm looking forward to see it in `master`! :smiley: 
Thanks @sorenmat

sorenmat commented:
@radeksimko glad to be able to help, and thanks for an awesome product !

I've renamed the command to validate and updated the documentation as requested.
I've also added a few more test to cover the checks defined in config.go.
Please note: There is no check in `config.Validate` that checks for multiple output's with the same name.
Is this an error / missing feature ? 

If it is I will be glad to do another PR with that fix :)

radeksimko commented:
> There is no check in config.Validate that checks for multiple output's with the same name.

Good catch! I just opened a new issue for this: https://github.com/hashicorp/terraform/issues/5052
If you want to send another PR to fix that, that's fine, I believe this is out of scope for this PR though.

I think this is good to go! :ship:
